crates/events/src/tracer.rs

```diff
@@ -0,0 +1,301 @@
+pub struct EventTracer<W> {
+    wrapped: W,
+    pkg_name: &'static str,
+    pkg_version: &'static str,
+}
+
+impl<W> EventTracer<W> {
+    /// Build a new [`EventTracer`].
+    pub const fn trace(wrapped: W) -> Self {
+        Self {
+            wrapped,
+            pkg_name: env!("CARGO_PKG_NAME"),
+            pkg_version: env!("CARGO_PKG_VERSION"),
+        }
+    }
+}
+
+impl<EE, W> EmitEvents for EventTracer<W>
+where
+    EE: EmitEvents,
+    W: DerefMut<Target = EE>,
+{
+    #[track_caller]
+    fn emit<E>(&mut self, event: E)
+    where
+        E: EventToEmit,
+    {
+        let caller = Location::caller();
+
+        self.wrapped.emit(event.with(EventOrigin(EventTrace {
+            pkg_name: Cow::Borrowed(self.pkg_name),
+            pkg_version: Cow::Borrowed(self.pkg_version),
+            file: Cow::Borrowed(caller.file()),
+            line: caller.line(),
+            column: caller.column(),
+        })));
+    }
+
+    #[track_caller]
+    fn emit_many<B, E>(&mut self, event_batch: B)
+    where
+        B: IntoIterator<Item = E>,
+        E: EventToEmit,
+    {
+        let caller = Location::caller();
+
+        self.wrapped.emit_many(event_batch.into_iter().map(|event| {
+            event.with(EventOrigin(EventTrace {
+                pkg_name: Cow::Borrowed(self.pkg_name),
+                pkg_version: Cow::Borrowed(self.pkg_version),
+                file: Cow::Borrowed(caller.file()),
+                line: caller.line(),
+                column: caller.column(),
+            }))
+        }));
+    }
+}
+
+/// Extend an [`Event`](super::Event) with data pertaining to its origin in
+/// source code.
+pub struct EventOrigin<'a>(pub EventTrace<'a>);
+
+impl<'a> EventAttributeEntry<'a> for EventOrigin<'a> {
+    type Value = EventTrace<'a>;
+    type ValueOwned = EventTrace<'static>;
+
+    const KEY: &'static str = "event-origin";
+
+    fn into_value(self) -> Self::Value {
+        self.0
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{Event, EventLevel, EventTypeBuilder};
+
+    const fn dummy_trace() -> EventTrace<'static> {
+        EventTrace {
+            pkg_name: Cow::Borrowed("pkg"),
+            pkg_version: Cow::Borrowed("ver"),
+            file: Cow::Borrowed("src/file.rs"),
+            line: 1,
+            column: 2,
+        }
+    }
+
+    #[test]
+    fn test_event_trace_emit_event() {
+        let (event, start_line, end_line) = {
+            let ev = Event::new(
+                EventTypeBuilder::new_with_type("test").build(),
+                EventLevel::Tx,
+            );
+            let mut events = Vec::with_capacity(1);
+
+            const START_LINE: u32 = line!();
+            emit_event(ev, &mut events);
+            const END_LINE: u32 = line!();
+
+            (events.pop().unwrap(), START_LINE, END_LINE)
+        };
+
+        let trace = event.read_attribute::<EventOrigin>().unwrap();
+
+        assert!(trace.line > start_line && trace.line < end_line);
```

Comment:
nit: maybe `trace.line == start_line + 1 && trace.line == END_LINE - 1`?

Reply:
doesn't matter to be precise, we only care to check if the event was emitted within a certain region of the source code